lib/Dialect/ESI/runtime/python/esiaccel/cosim/simulator.py

```diff
@@ -161,14 +223,13 @@ def run_proc(self, gui: bool = False) -> SimProcess:
         # Slow the simulation down to one tick per millisecond.
         simEnv["DEBUG_PERIOD"] = "1"
     rcmd = self.run_command(gui)
-    simProc = subprocess.Popen(self.run_command(gui),
-                               stdout=simStdout,
-                               stderr=simStderr,
-                               env=simEnv,
-                               cwd=self.run_dir,
-                               preexec_fn=os.setsid)
-    simStderr.close()
-    simStdout.close()
+    rcmd = self.run_command(gui)
+    simProc = self._start_process_with_callbacks(rcmd,

```

Comment:
Where do you join with the two threads you started and didn't join with because `wait=False`?

Reply:
There is another race condition wherein the simulator process has stopped but the threads have not finished processing the input. esi-cosim then exits without the callbacks having processed all of the inputs. This is actually plausible if the simulation outputs a bunch of text at the very end of the sim.

The other reason to `join()` on all threads is that we probably want to add a feature wherein the callbacks detect a failure. We want this to change the exit code but don't want the callbacks to terminate immediately by calling `sys.exit(1)`.